cognite/client/_api/datapoints.py

```diff
@@ -1,117 +1,969 @@
+            (similar to how slicing works with non-unique indexes on Pandas DataFrames):
+
+                >>> dps_lst = client.time_series.data.retrieve(
+                ...     id=[
+                ...         42, 43, 44, 45,
+                ...         {"id": 350, "start": datetime(1907, 10, 14, tzinfo=utc), "end": datetime(1907, 11, 6, tzinfo=utc)},
+                ...         {"id": 350, "start": datetime(1929, 9, 4, tzinfo=utc), "end": datetime(1929, 11, 13, tzinfo=utc)},
+                ...     ])
+                >>> ts_44 = dps_lst.get(id=44)  # Single `Datapoints` object
+                >>> ts_350_lst = dps_lst.get(id=350)  # List of two `Datapoints` objects
+
+            The last example showcases the great flexibility of the `retrieve` endpoint, with a very custom query. If you also want to
+            specify multiple values for `ignore_unknown_ids`, you'll need to use the `.query` endpoint.
+
+                >>> ts1 = 1337
+                >>> ts2 = {
+                ...     "id": 42,
+                ...     "start": -12345,  # Overrides `start` argument below
+                ...     "end": "1h-ago",
+                ...     "limit": 1000,  # Overrides `limit` argument below
+                ...     "include_outside_points": True
+                ... }
+                >>> ts3 = {
+                ...     "id": 11,
+                ...     "end": "1h-ago",
+                ...     "aggregates": ["max"],
+                ...     "granularity": "42h",
+                ...     "include_outside_points": False
+                ... }
+                >>> dps = client.time_series.data.retrieve(
+                ...    id=[ts1, ts2, ts3], start="2w-ago", limit=None
+                ... )
+        """
+        query = DatapointsQuery(
+            start=start,
+            end=end,
+            id=id,
+            external_id=external_id,
+            aggregates=aggregates,
+            granularity=granularity,
+            limit=limit,
+            include_outside_points=include_outside_points,
+            ignore_unknown_ids=ignore_unknown_ids,
+        )
+        fetcher = dps_fetch_selector(self, user_queries=[query])
+        dps_list = fetcher.fetch_all_datapoints(use_numpy=False)
+        if not query.is_single_identifier:
+            return dps_list
+        elif not dps_list and ignore_unknown_ids:
+            return None
+        return dps_list[0]
+
+    def retrieve_arrays(
+        self,
+        *,
+        id: Optional[DatapointsIdTypes] = None,
+        external_id: Optional[DatapointsExternalIdTypes] = None,
+        start: Union[int, str, datetime, None] = None,
+        end: Union[int, str, datetime, None] = None,
+        aggregates: Optional[List[str]] = None,
+        granularity: Optional[str] = None,
+        limit: Optional[int] = None,
+        include_outside_points: bool = False,
+        ignore_unknown_ids: bool = False,
+    ) -> Union[None, DatapointsArray, DatapointsArrayList]:
```

Comment:
Seems to me like DatapointsArray and DatapointsArrayList have the exact same interface as Datapoints and DatapointsList, respectively. I think we can just make a common protocols `Datapoints` and `DatapointsList`, and then have separate implementations for the numpy and non-numpy versions. Then, we won't need a separate `retrieve_arrays`, we just make the enable the optimizations behind the scenes if numpy is installed.

Reply:
Discussed and resolved on slack: https://cognitedata.slack.com/archives/C03PBAB4KUJ/p1664797575423249